// @sentry/aws-serverless/cjs/index.js
Object.defineProperty(exports, '__esModule', { value: true });

const node = require('@sentry/node');
const core = require('@sentry/core');
const aws = require('./integration/aws.js');
const awslambda = require('./integration/awslambda.js');
const sdk = require('./sdk.js');



exports.DEFAULT_USER_INCLUDES = node.DEFAULT_USER_INCLUDES;
exports.NodeClient = node.NodeClient;
exports.SDK_VERSION = node.SDK_VERSION;
exports.SEMANTIC_ATTRIBUTE_SENTRY_OP = node.SEMANTIC_ATTRIBUTE_SENTRY_OP;
exports.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN = node.SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE = node.SEMANTIC_ATTRIBUTE_SENTRY_SAMPLE_RATE;
exports.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE = node.SEMANTIC_ATTRIBUTE_SENTRY_SOURCE;
exports.Scope = node.Scope;
exports.addBreadcrumb = node.addBreadcrumb;
exports.addEventProcessor = node.addEventProcessor;
exports.addIntegration = node.addIntegration;
exports.addOpenTelemetryInstrumentation = node.addOpenTelemetryInstrumentation;
exports.addRequestDataToEvent = node.addRequestDataToEvent;
exports.anrIntegration = node.anrIntegration;
exports.captureCheckIn = node.captureCheckIn;
exports.captureEvent = node.captureEvent;
exports.captureException = node.captureException;
exports.captureFeedback = node.captureFeedback;
exports.captureMessage = node.captureMessage;
exports.captureSession = node.captureSession;
exports.close = node.close;
exports.connectIntegration = node.connectIntegration;
exports.consoleIntegration = node.consoleIntegration;
exports.contextLinesIntegration = node.contextLinesIntegration;
exports.continueTrace = node.continueTrace;
exports.createGetModuleFromFilename = node.createGetModuleFromFilename;
exports.createTransport = node.createTransport;
exports.cron = node.cron;
exports.defaultStackParser = node.defaultStackParser;
exports.endSession = node.endSession;
exports.expressErrorHandler = node.expressErrorHandler;
exports.expressIntegration = node.expressIntegration;
exports.extractRequestData = node.extractRequestData;
exports.fastifyIntegration = node.fastifyIntegration;
exports.flush = node.flush;
exports.functionToStringIntegration = node.functionToStringIntegration;
exports.getActiveSpan = node.getActiveSpan;
exports.getAutoPerformanceIntegrations = node.getAutoPerformanceIntegrations;
exports.getClient = node.getClient;
exports.getCurrentHub = node.getCurrentHub;
exports.getCurrentScope = node.getCurrentScope;
exports.getGlobalScope = node.getGlobalScope;
exports.getIsolationScope = node.getIsolationScope;
exports.getRootSpan = node.getRootSpan;
exports.getSentryRelease = node.getSentryRelease;
exports.getSpanDescendants = node.getSpanDescendants;
exports.getSpanStatusFromHttpCode = node.getSpanStatusFromHttpCode;
exports.graphqlIntegration = node.graphqlIntegration;
exports.hapiIntegration = node.hapiIntegration;
exports.httpIntegration = node.httpIntegration;
exports.inboundFiltersIntegration = node.inboundFiltersIntegration;
exports.initOpenTelemetry = node.initOpenTelemetry;
exports.isInitialized = node.isInitialized;
exports.koaIntegration = node.koaIntegration;
exports.linkedErrorsIntegration = node.linkedErrorsIntegration;
exports.localVariablesIntegration = node.localVariablesIntegration;
exports.makeNodeTransport = node.makeNodeTransport;
exports.metrics = node.metrics;
exports.modulesIntegration = node.modulesIntegration;
exports.mongoIntegration = node.mongoIntegration;
exports.mongooseIntegration = node.mongooseIntegration;
exports.mysql2Integration = node.mysql2Integration;
exports.mysqlIntegration = node.mysqlIntegration;
exports.nativeNodeFetchIntegration = node.nativeNodeFetchIntegration;
exports.nestIntegration = node.nestIntegration;
exports.nodeContextIntegration = node.nodeContextIntegration;
exports.onUncaughtExceptionIntegration = node.onUncaughtExceptionIntegration;
exports.onUnhandledRejectionIntegration = node.onUnhandledRejectionIntegration;
exports.parameterize = node.parameterize;
exports.postgresIntegration = node.postgresIntegration;
exports.prismaIntegration = node.prismaIntegration;
exports.redisIntegration = node.redisIntegration;
exports.requestDataIntegration = node.requestDataIntegration;
exports.setContext = node.setContext;
exports.setCurrentClient = node.setCurrentClient;
exports.setExtra = node.setExtra;
exports.setExtras = node.setExtras;
exports.setHttpStatus = node.setHttpStatus;
exports.setMeasurement = node.setMeasurement;
exports.setTag = node.setTag;
exports.setTags = node.setTags;
exports.setUser = node.setUser;
exports.setupConnectErrorHandler = node.setupConnectErrorHandler;
exports.setupExpressErrorHandler = node.setupExpressErrorHandler;
exports.setupHapiErrorHandler = node.setupHapiErrorHandler;
exports.setupKoaErrorHandler = node.setupKoaErrorHandler;
exports.setupNestErrorHandler = node.setupNestErrorHandler;
exports.spanToBaggageHeader = node.spanToBaggageHeader;
exports.spanToJSON = node.spanToJSON;
exports.spanToTraceHeader = node.spanToTraceHeader;
exports.spotlightIntegration = node.spotlightIntegration;
exports.startInactiveSpan = node.startInactiveSpan;
exports.startSession = node.startSession;
exports.startSpan = node.startSpan;
exports.startSpanManual = node.startSpanManual;
exports.trpcMiddleware = node.trpcMiddleware;
exports.withActiveSpan = node.withActiveSpan;
exports.withIsolationScope = node.withIsolationScope;
exports.withMonitor = node.withMonitor;
exports.withScope = node.withScope;
exports.zodErrorsIntegration = node.zodErrorsIntegration;
exports.captureConsoleIntegration = core.captureConsoleIntegration;
exports.debugIntegration = core.debugIntegration;
exports.dedupeIntegration = core.dedupeIntegration;
exports.extraErrorDataIntegration = core.extraErrorDataIntegration;
exports.rewriteFramesIntegration = core.rewriteFramesIntegration;
exports.sessionTimingIntegration = core.sessionTimingIntegration;
exports.awsIntegration = aws.awsIntegration;
exports.awsLambdaIntegration = awslambda.awsLambdaIntegration;
exports.getDefaultIntegrations = sdk.getDefaultIntegrations;
exports.init = sdk.init;
exports.tryPatchHandler = sdk.tryPatchHandler;
exports.wrapHandler = sdk.wrapHandler;
